sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1516,6 +1553,53 @@ pi_result _pi_context::getAvailableCommandList(
   return pi_result;
 }
 
+_pi_queue::pi_queue_group_t &_pi_queue::getQueueGroup(bool UseCopyEngine) {
+  // Check if thread-specifc immediate commandlists are requested.
+  if (Device->useImmediateCommandLists() == PerThreadPerQueue) {
+    // Thread id is used to create separate imm cmdlists per thread.
+    auto TID = std::this_thread::get_id();
+    if (UseCopyEngine) {
+      auto InitialGroup = CopyQueueGroupsByTID.begin()->second;
+      std::pair<std::unordered_map<std::thread::id, pi_queue_group_t>::iterator,
+                bool>
+          Result = CopyQueueGroupsByTID.insert({TID, InitialGroup});
+      // If an entry for this thread exists, use it.
+      if (!Result.second) {
+        return Result.first->second;
+      }
+      auto &CopyQueueGroupRef = Result.first->second;
+      // Create space to hold immediate commandlists. They will be created on
+      // demand.
+      CopyQueueGroupRef.ImmCmdLists = std::vector<pi_command_list_ptr_t>(
+          InitialGroup.ZeQueues.size(), CommandListMap.end());
+      return CopyQueueGroupRef;
+    } else {
+      auto InitialGroup = ComputeQueueGroupsByTID.begin()->second;
+      std::pair<std::unordered_map<std::thread::id, pi_queue_group_t>::iterator,
+                bool>
+          Result = ComputeQueueGroupsByTID.insert({TID, InitialGroup});
+      // If an entry for this thread exists, use it.
+      if (!Result.second)
+        return Result.first->second;
+
+      auto &ComputeQueueGroupRef = Result.first->second;
+      // Create space to hold immediate commandlists. They will be created on
+      // demand.
+      ComputeQueueGroupRef.ImmCmdLists = std::vector<pi_command_list_ptr_t>(
+          InitialGroup.ZeQueues.size(), CommandListMap.end());
+      return ComputeQueueGroupRef;
+    }
```

Comment:
the if/else do the same thing on different objects, please combine

Reply:
Yes, even further code reduction was possible and has been done.